providers/configcat/src/main/java/dev/openfeature/contrib/providers/configcat/ConfigCatProvider.java

```diff
@@ -0,0 +1,205 @@
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Provider implementation for ConfigCat.
+ */
+@Slf4j
+public class ConfigCatProvider extends EventProvider {
+
+    @Getter
+    private static final String NAME = "ConfigCat";
+
+    public static final String PROVIDER_NOT_YET_INITIALIZED = "provider not yet initialized";
+    public static final String UNKNOWN_ERROR = "unknown error";
+
+    private ConfigCatProviderConfig configCatProviderConfig;
+
+    @Getter
+    private ConfigCatClient configCatClient;
+
+    @Getter
+    private ProviderState state = ProviderState.NOT_READY;
+
+    private AtomicBoolean isInitialized = new AtomicBoolean(false);
+
+    /**
+     * Constructor.
+     * @param configCatProviderConfig configCatProvider Config
+     */
+    public ConfigCatProvider(ConfigCatProviderConfig configCatProviderConfig) {
+        this.configCatProviderConfig = configCatProviderConfig;
+    }
+
+    /**
+     * Initialize the provider.
+     * @param evaluationContext evaluation context
+     * @throws Exception on error
+     */
+    @Override
+    public void initialize(EvaluationContext evaluationContext) throws Exception {
+        boolean initialized = isInitialized.getAndSet(true);
+        if (initialized) {
+            throw new GeneralError("already initialized");
+        }
+        super.initialize(evaluationContext);
+        configCatClient = ConfigCatClient.get(configCatProviderConfig.getSdkKey(),
+            configCatProviderConfig.getOptions());
+        configCatProviderConfig.postInit();
+        state = ProviderState.READY;
+        log.info("finished initializing provider, state: {}", state);
+
+        configCatClient.getHooks().addOnClientReady(() -> {
+            state = ProviderState.READY;
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .message("provider ready")
+                .build();
+            emitProviderReady(providerEventDetails);
+        });
+
+        configCatClient.getHooks().addOnConfigChanged(map -> {
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .flagsChanged(new ArrayList<>(map.keySet()))
+                .message("config changed")
+                .build();
+            emitProviderReady(providerEventDetails);
+        });
+
+        configCatClient.getHooks().addOnError(errorMessage -> {
+            ProviderEventDetails providerEventDetails = ProviderEventDetails.builder()
+                .message(errorMessage)
+                .build();
+            emitProviderError(providerEventDetails);
+        });
+    }
+
+    @Override
+    public Metadata getMetadata() {
+        return () -> NAME;
+    }
+
+    @Override
+    public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue, EvaluationContext ctx) {
+        if (!ProviderState.READY.equals(state)) {
```

Comment:
The ConfigCat SDK has a pretty good way to handle the client initialization delays. 
In auto polling mode, you can set a maxInitWaitTimeSeconds (defaults to 5 seconds).
If you call the getValue or getValueDetails right after the configcatclient initialization, and the client is not ready yet, the client will wait up to the maxInitWaitTimeSeconds to be ready and to provide a valid value. This is a pretty good safety net and I think it would be better to rely on this behaviour instead of throwing errors. 
So I'd remove all these throw errors from the beginning of these methods.

Reply:
It is not expected to get to these, there are still corner cases from invalid usage which gets to the flow of these checks:

- Constructing a provider directly and evaluate before initialization.
- Evaluating after shutdown.

Then prefer to leave them.

These cases are covered at the [test](https://github.com/open-feature/java-sdk-contrib/pull/521/files#diff-9e3dc32d15e260688a9aa99b1f5a361e71d54ababbf2940570dfacc7ec58ba09R128).